### PR TITLE
[RFC] Add support for TA profiling (gprof)

### PR DESCRIPTION
**Pull request by jforissier**
Enable with: CFG_TA_GPROF_SUPPORT=y.

Signed-off-by: Jerome Forissier jerome.forissier@linaro.org

### CODE REVIEW

File: tee-supplicant/src/gprof.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	uint32_t offset;
+	uint32_t res;
+	uint32_t id; /* */
+	/* len bytes of data follow */
+};
+
+int gprof_process(void *cmd, size_t cmd_size)
+{
+	struct gprof_rpc *rpc = cmd;
+	char *buf = (char *)(rpc + 1);
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: Please use a more modern approach to pass RPC arguments like the one in https://github.com/OP-TEE/optee_client/pull/58

File: tee-supplicant/src/gprof.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			u->clockSeqAndNode[6], u->clockSeqAndNode[7],
+			vers);
+	        if ((n < 0) || (n >= (int)sizeof(path)))
+			break;
+		flags = O_WRONLY;
+		if (!rpc->offset)
+			flags |= O_CREAT | O_EXCL;
+		fd = open(path, flags, 0600);
+		if (fd >= 0) {
+			st = lseek(fd, rpc->offset, SEEK_SET);
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: use pwrite(2) below instead?

File: tee-supplicant/src/gprof.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <tee_client_api.h>
+#include "gprof.h"
+
+struct gprof_rpc {
+	TEEC_UUID ta_uuid;
+	uint32_t len;
+	uint32_t offset;
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: Why do we need an offset, wouldn't append only be more useful?

File: tee-supplicant/src/gprof.c (hunk 1 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: I will.

File: tee-supplicant/src/gprof.c (hunk 2 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: Yep

File: tee-supplicant/src/gprof.c [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	TEEC_UUID *u;
+	size_t len;
+	void *buf;
+	int id;
+	int flags;
+	int fd = -1;
+	int st;
+	int n;
+
+	if (recv->num_params != 2 ||
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: Please document this interface `optee_msg.h` or some other suitable header file.

File: tee-supplicant/src/gprof.c [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+			"%s.out",
+			u->timeLow, u->timeMid, u->timeHiAndVersion,
+			u->clockSeqAndNode[0], u->clockSeqAndNode[1],
+			u->clockSeqAndNode[2], u->clockSeqAndNode[3],
+			u->clockSeqAndNode[4], u->clockSeqAndNode[5],
+			u->clockSeqAndNode[6], u->clockSeqAndNode[7],
+			vers);
+	        if ((n < 0) || (n >= (int)sizeof(path)))
+			break;
+		fd = open(path, flags, 0600);
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: Please use `S_IWUSR | S_IRUSR` instead of the magic number.

File: tee-supplicant/src/gprof.c [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		id = 1;
+	}
+
+	for (;;) {
+		if (id > 1) {
+			/*
+			 * id == 1 is file 0 (no suffix), id == 2 is file .1
+			 * etc.
+			 */
+			sprintf(vers, ".%d", id - 1);
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: `sprintf()` is not safe, please use `snprintf()` instead.

File: tee-supplicant/src/gprof.c [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	buf = u + 1;
+
+	if (id < 0 || id > 100)
+		return TEEC_ERROR_BAD_PARAMETERS;
+
+	flags = O_APPEND | O_WRONLY;
+	if (!id) {
+		/* id == 0 means create file and allocate id */
+		flags |= O_CREAT | O_EXCL;
+		id = 1;
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: This isn't exactly allocated.

File: tee-supplicant/src/gprof.c [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		return TEEC_ERROR_BAD_PARAMETERS;
+
+	id = params[0].u.value.a;
+	if (params[1].u.memref.size < sizeof(TEEC_UUID))
+		return TEEC_ERROR_BAD_PARAMETERS;
+	len = params[1].u.memref.size - sizeof(TEEC_UUID);
+	u = tee_supp_param_to_va(params + 1);
+	if (!u)
+		return TEEC_ERROR_BAD_PARAMETERS;
+	buf = u + 1;
```
Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: Why not use a `param[2]` for this buffer instead?

File: tee-supplicant/src/gprof.c (hunk 5 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: Well, the symbolic names are not easy to read, and "everyone knows what 0600 is": https://marc.info/?l=linux-kernel&m=147017153702150&w=2

Comment (jenswi-linaro) [inline tee-supplicant/src/gprof.c]: Haha, do as you please then. :-)

File: tee-supplicant/src/gprof.c (hunk 4 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: ok

File: tee-supplicant/src/gprof.c (hunk 8 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: That's what I did initially and got a null pointer from `tee_supp_param_to_va(params + 2)`. I'll try harder ;)

File: tee-supplicant/src/gprof.c (hunk 7 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: ok

File: tee-supplicant/src/gprof.c (hunk 6 above)
Comment (jforissier) [inline tee-supplicant/src/gprof.c]: yes

**Comment by jenswi-linaro** [comment]
Looks good, now only the protocol used by `RPC_CMD_GPROF` need to be described somewhere. Perhaps `optee_msg_fs.h` should be renamed to `optee_msg_supplicant.h` and we could describe the protocol used by supplicant there.

**Comment by jforissier** [comment]
+1 for `optee_msg_supplicant.h`. In both optee_os and optee_client I suppose? That is, should I rename:
```
optee_os/core/include/optee_msg_fs.h -> optee_os/core/include/optee_msg_supplicant.h
optee_client/tee-supplicant/src/optee_msg_fs.h -> optee_client/tee-supplicant/src/optee_msg_supplicant.h
```
?
(These two files should be identical)

**Comment by jenswi-linaro** [comment]
Yes. Another advantage here is that the `optee_msg.h` file that we sync with the Linux kernel doesn't need updates that often.

**Comment by jforissier** [comment]
OK, I'll do that.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
Rebased and tag applied. To be merged at the same time as the OP-TEE OS pull request.

**Comment by jenswi-linaro** [comment]
Anything we're waiting for here?

**Pull request merged**